clippy_lints/src/utils/attrs.rs

```diff
@@ -21,25 +21,27 @@ pub const BUILTIN_ATTRIBUTES: &[(&str, DeprecationStatus)] = &[
         DeprecationStatus::Replaced("cognitive_complexity"),
     ),
     ("dump", DeprecationStatus::None),
+    ("msrv", DeprecationStatus::None),
 ];
 
-pub struct LimitStack {
-    stack: Vec<u64>,
+#[derive(Default, Clone)]
+pub struct LimitStack<T> {
```

Comment:
@flip1995 should we call this something else since it's not just being used by the `cognitive_complexity` lint anymore?

Reply:
I don't think we need a stack for this. Setting MSRV for a single function/module, instead of a whole crate doesn't really make sense to me.

Best case would be, if we could detect if this is a inner `#![...]` or outer `#[...]` attribute and error out in the later case. With this, we only need to set the MSRV once and don't have to check for the attribute on every item.

Reply:
@flip1995 that makes sense. I think we can check the [`AttrStyle`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast/ast/enum.AttrStyle.html) to figure out if this is an inner or outer attribute. In case the msrv is also specified in `clippy.toml`, will this override that value?

Reply:
@flip1995 setting the `msrv` as an inner attribute throws the following error. Should it be an outer one instead?

```rust
error[E0658]: non-builtin inner attributes are unstable
  --> $DIR/manual_strip.rs:2:1
   |
LL | #![clippy::msrv = "1.45.0"]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #54726 <https://github.com/rust-lang/rust/issues/54726> for more information
   = help: add `#![feature(custom_inner_attributes)]` to the crate attributes to enable

error: aborting due to previous error

For more information about this error, try `rustc --explain E0658`
```

Reply:
> I think we can check the [`AttrStyle`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast/ast/enum.AttrStyle.html) to figure out if this is an inner or outer attribute. In case the msrv is also specified in `clippy.toml`, will this override that value?

Yes, this should override the value set in `clippy.toml`.

Just checking the `AttrStyle` won't be enough, since you could set inner attributes in modules. But this should be good enough for now.



> setting the `msrv` as an inner attribute throws the following error. Should it be an outer one instead?

Oh, that's too bad. Crates that have a MSRV most likely use a stable toolchain. This means, they won't be able to set this attribute, so it would be quite useless (for now). On the other hand, only allowing outer attributes is also useless. 

I would add this anyway, since this should be the way to do it (as long as cargo doesn't support MSRV). And then the question is, if this feature is stable first or cargo support for MSRV.

Reply:
@flip1995 since we're adding this as an inner attribute, any idea how to set the `msrv`  in the tests for the affected lints since the default is 1.0.0?

Reply:
>  any idea how to set the `msrv` in the tests for the affected lints since the default is 1.0.0?

I don't think we have to do this. We always test Clippy with the latest nightly, so we always meet the MSRV for those lints in Clippy.

Reply:
@flip1995 don't think this should be a problem now as we're defaulting to `None` instead of `1.0.0`